lang/src/org/partiql/lang/syntax/SqlParser.kt

```diff
@@ -1878,15 +1875,134 @@ class SqlParser(private val ion: IonSystem) : Parser {
+                var children : List<ParseNode>
+                var joinToken : Token? = delim.token
+
+                rem = rem.tail
+
+                if (hasOnClause) {
+                    // Explicit join
+                    if (rem.head?.type == LEFT_PAREN) {
+                        // Starts with a left paren. Could indicate subquery/literal or indicate higher precedence
+                        val isSubqueryOrLiteral = rem.tail.head?.type == LITERAL || rem.tail.head?.keywordText == "select"
+                        val parenClause = rem.parseFromSource(precedence, true)
+                        rem = parenClause.remaining
+
+                        // check for an ON-clause
+                        if (rem.head?.keywordText != "on") {
+                            rem.err("Expected 'ON'", PARSE_MALFORMED_JOIN)
+                        }
+
+                        val onClause = rem.tail.parseExpression(precedence)
+
+                        rem = onClause.remaining
+                        if (!isSubqueryOrLiteral) {
+                            children = listOf(parenClause, left, onClause)
+                        }
+
+                        else {
+                            children = listOf(left, parenClause, onClause)
+                        }
+                    }
+
+                    else {
+                        // Rest is just the right side of the clause
+                        val rightRef = rem.parseFromSource(precedence, false)
+                        rem = rightRef.remaining
+
+                        // check for an ON-clause
+                        if (rem.head?.keywordText != "on") {
+                            rem.err("Expected 'ON'", PARSE_MALFORMED_JOIN)
+                        }
+
+                        val onClause = rem.tail.parseExpression(precedence)
+
+                        rem = onClause.remaining
+
+                        children = listOf(left, rightRef, onClause)
+                    }
+                }
+
+                else {
+                    // For implicit joins
+                    val rightRef = rem.parseFromSource(precedence, false)
+                    rem = rightRef.remaining
+                    children = listOf(left, rightRef)
+                    if (delim.token?.type == COMMA) {
+                        joinToken = delim.token?.copy(
+                            type = KEYWORD,
+                            value = ion.newSymbol("cross_join")
+                        )
+                    }
+                }
+                left = ParseNode(FROM_SOURCE_JOIN, joinToken, children, rem)
+                delim = rem.parseJoinDelim()
+            }
+            return left
+        }
+        return child
+    }
+
+
+    private fun List<Token>.parseFromSourceList(precedence: Int = -1): ParseNode {
+        var rem = this
+        val child = rem.parseFromSource(precedence)
+        rem = child.remaining
+
+        return ParseNode(FROM_CLAUSE, null, listOf(child), rem)
```

Comment:
Use named arguments with `null` literals

Reply:
Let as is since can't use named arguments with positional args for a constructor. Also, there are a number of existing calls to ParseNode with `null` as an argument